### PR TITLE
enable dispatch stub for backend PrivateUse1

### PR DESCRIPTION
**Pull request by caizhi-mt**
When expanding the new backend of pytorch in the form of out ot tree, Privateuse1 will be reused. So we also need to support PrivateUse1 in the dispatch stub module

cc @bdhirsh

### CODE REVIEW

**Comment by caizhi-mt** [comment]
@bdhirsh Please help me review this commit, thank you

File: aten/src/ATen/native/DispatchStub.h
Comment (bdhirsh) [inline aten/src/ATen/native/DispatchStub.h:175]: nit: can you make it `set_privateuse1_dispatch_ptr` (for greppability)

**Comment by bdhirsh** [comment]
One comment: I don't think getting DispatchStub to work with externally backends, is **strictly** necessary, since for any ATen kernels that are implemented with it, you can always register your own dedicated kernel to the PrivateUse1 dispatch key for that operator.

Is the idea that you want to be able to re-use some kernels from ATen that are implemented in terms of dispatch stub? That sounds pretty useful, I just wanted to confirm.

If so, it would be good to add a test. Do you think you can add a DispatchStub handler in `test/cpp_extensions/open_registration_extension.cpp`, and exercise that logic in `test/test_cpp_extensions_open_device_registration.py`?

File: aten/src/ATen/native/DispatchStub.h
Comment (caizhi-mt) [reply aten/src/ATen/native/DispatchStub.h]: done

File: test/cpp_extensions/open_registration_extension.cpp
Comment (bdhirsh) [inline test/cpp_extensions/open_registration_extension.cpp:57]: it would be nice to be able to test that we actually ran this kernel (right now the "test" just calls `abs()` and checks that the outputs match).

One way to do it would be the way it's done in `custom_add_Tensor` below - add a global counter for whether our custom abs kernel was called, that we increment inside of the kernel. Then expose a `custom_abs_called` helper to python that you can assert in the python code.

Comment (caizhi-mt) [reply test/cpp_extensions/open_registration_extension.cpp:47]: Thank you for the suggestion. I have modified the unit test according to your suggestion, added `custom_abs_called` and `abs_counter` in test.

**Comment by bdhirsh** [comment]
Can you take a look at the failing CI?

It looks like the custom cpp extension is failing to be compiled:
```
2023-05-01T10:03:59.2852263Z  const at::Tensor& custom_resize_(const at::Tensor& self, at::IntArrayRef size,
2023-05-01T10:03:59.2852610Z                    ^~~~~~~~~~~~~~
2023-05-01T10:03:59.2853410Z /var/lib/jenkins/workspace/test/cpp_extensions/open_registration_extension.cpp:63:19: note: ‘const at::Tensor& custom_resize_(const at::Tensor&, c10::IntArrayRef, c10::optional<c10::MemoryFormat>)’ previously defined here
2023-05-01T10:03:59.2854072Z  const at::Tensor& custom_resize_(
2023-05-01T10:03:59.2854350Z                    ^~~~~~~~~~~~~~
2023-05-01T10:03:59.2854649Z ninja: build stopped: subcommand failed.
```

File: test/cpp_extensions/open_registration_extension.cpp
Comment (bdhirsh) [inline test/cpp_extensions/open_registration_extension.cpp:68]: it looks like it's failing because `custom_resize_()` is already defined further down.

You can just use the existing one instead of defining this new one.

Comment (caizhi-mt) [reply test/cpp_extensions/open_registration_extension.cpp]: ok, I have updated the code

Comment (caizhi-mt) [reply test/cpp_extensions/open_registration_extension.cpp]: all checks have passed, could this PR be merged?

**Comment by caizhi-mt** [comment]
This PR need to be approved by an authorized maintainer before merge. Could you approve this PR? @bdhirsh

**Comment by caizhi-mt** [comment]
@bdhirsh  Could you approve this PR? Thank you

**Comment by caizhi-mt** [comment]
@ezyang Could you review and approve this PR? Thank you

**Comment by ezyang** [comment]
Sorry, y'all have already gone a review on this, but I want to put the breaks on this. The dispatch stub API is not intended for external users, and for the most part we've expunged all the cases where we registered kernels that call into stubs so that they don't CompositeExplicitAutograd anymore. Concretely, what operators do you think you need this for?

**Comment by caizhi-mt** [comment]
Actually I want to extend a new hardware backend based on "PrivateUse1" key. And I want to use the dispatchstub mechanism to call my kernels. such as:  https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/UnaryOps.cpp#L446
This is why I commit this PR。

**Comment by caizhi-mt** [comment]
yes, I actually used TensorIterator.

1. I am doing some cuda-compatible work using Privateuse1 key base on our hardware which will reuse cuda kernels.
2. We will refer to cuda kernels to implement our own kernels using TensorIterator

PrivateUse1 key is just like "CPU" key, "CUDA" key. I think it is necessary to support privateuse1 in dispatchstub.

**Review by ezyang — APPROVED** [review]
ok fine

**Pull request closed without merging**
